### PR TITLE
add types in bolts/callbacks

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Part of #434 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ydcjeff** [comment]
I need help with the failed mypy check.

**Comment by ydcjeff** [comment]
matplotlib is needed for adding types in confused logits callbacks. how shall we do it?

File: pl_bolts/callbacks/byol_updates.py [hunk 1]
```diff
@@ -36,7 +39,8 @@ def __init__(self, initial_tau: float = 0.996):
         self.initial_tau = initial_tau
         self.current_tau = initial_tau
 
-    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
+    def on_train_batch_end(self, trainer: Trainer, pl_module: LightningModule, outputs: Sequence,
```
Comment (Borda) [inline pl_bolts/callbacks/byol_updates.py:42]: lest spit all these multi-line to argument per line
```suggestion
    def on_train_batch_end(self, trainer: Trainer, pl_module: LightningModule, outputs: Sequence,
```

File: pl_bolts/callbacks/data_monitor.py [hunk 2]
```diff
@@ -189,7 +189,7 @@ def _register_hook(self, module_name: str, module: nn.Module) -> RemovableHandle
             else self.GROUP_NAME_OUTPUT
         )
 
-        def hook(_, inp, out):
+        def hook(_: nn.Module, inp: Sequence, out: Sequence) -> None:
```
Comment (Borda) [inline pl_bolts/callbacks/data_monitor.py:192]:
```suggestion
        def hook(_: Module, inp: Sequence, out: Sequence) -> None:
```

File: pl_bolts/callbacks/data_monitor.py [hunk 3]
```diff
@@ -219,7 +219,8 @@ def __init__(self, log_every_n_steps: int = None):
         """
         super().__init__(log_every_n_steps=log_every_n_steps)
 
-    def on_train_batch_start(self, trainer, pl_module, batch, *args, **kwargs):
+    def on_train_batch_start(self, trainer: Trainer, pl_module: LightningModule,
```
Comment (Borda) [inline pl_bolts/callbacks/data_monitor.py:222]: formating :]

File: setup.cfg [hunk 4]
```diff
@@ -67,11 +67,10 @@ description-file = README.md
 # untyped test functions (using `--check-untyped-defs`) is still
 # high-value because it helps test the typing.
 files = pl_bolts, tests
+pretty = True
 disallow_untyped_defs = True
 ignore_missing_imports = True
 
-[mypy-pl_bolts.callbacks.*]
```
Comment (Borda) [inline setup.cfg:73]: nice! 🎉

**Comment by akihironitta** [comment]
@ydcjeff 
> matplotlib is needed for adding types in confused logits callbacks. how shall we do it?

I'm adding `matplotlib` to the (optional) requirements in #450, so after having it merged, it should work without any issue I suppose :]

**Comment by Borda** [comment]
@ydcjeff can we resolve conflits?

File: pl_bolts/callbacks/data_monitor.py [hunk 5]
```diff
@@ -219,7 +220,13 @@ def __init__(self, log_every_n_steps: int = None):
...
         super().__init__(log_every_n_steps=log_every_n_steps)
 
-    def on_train_batch_start(self, trainer, pl_module, batch, *args, **kwargs):
+    def on_train_batch_start(self,
+                             trainer: Trainer,
+                             pl_module: LightningModule,
+                             batch: Sequence,
+                             *args: int,
+                             **kwargs: int
+                             ) -> None:
```
Comment (Borda) [inline pl_bolts/callbacks/data_monitor.py:229]:
```suggestion
    def on_train_batch_start(
            self,
            trainer: Trainer,
            pl_module: LightningModule,
            batch: Sequence,
            *args: int,
            **kwargs: int,
    ) -> None:
```

File: pl_bolts/callbacks/variational.py [hunk 6]
```diff
@@ -19,6 +22,14 @@ class LatentDimInterpolator(Callback):
...
     Default interpolates between [-5, 5] (-5, -4, -3, ..., 3, 4, 5)
 
+    Args:
+        interpolate_epoch_interval: default 20
+        range_start: default -5
+        range_end: default 5
+        steps: number of step between start and end
+        num_samples: default 2
+        normalize: default True (change image to (0, 1) range)
+
```
Comment (Borda) [inline pl_bolts/callbacks/variational.py:32]:
```suggestion
```

File: pl_bolts/callbacks/variational.py [hunk 7]
```diff
@@ -35,15 +46,6 @@ def __init__(
         num_samples: int = 2,
         normalize: bool = True,
     ):
```
Comment (Borda) [inline pl_bolts/callbacks/variational.py:48]:
```suggestion
    ):
        """	
        Args:	
            interpolate_epoch_interval: default 20	
            range_start: default -5	
            range_end: default 5	
            steps: number of step between start and end	
            num_samples: default 2	
            normalize: default True (change image to (0, 1) range)	
        """
```
lets keep it in init

File: pl_bolts/callbacks/vision/confused_logit.py [hunk 8]
```diff
@@ -102,7 +106,13 @@ def training_step(...):
...
                 pl_module.train()
 
-    def _plot(self, confusing_x, confusing_y, trainer, model, mask_idxs):
+    def _plot(self,
+              confusing_x: Tensor,
+              confusing_y: Tensor,
+              trainer: Trainer,
+              model: LightningModule,
+              mask_idxs: Tensor
+              ) -> None:
```
Comment (Borda) [inline pl_bolts/callbacks/vision/confused_logit.py:115]:
```suggestion
    def _plot(
            self,
            confusing_x: Tensor,
            confusing_y: Tensor,
            trainer: Trainer,
            model: LightningModule,
            mask_idxs: Tensor,
    ) -> None:
```

**Review by Borda — APPROVED** [review]

File: pl_bolts/callbacks/byol_updates.py [hunk 9]
```diff
@@ -36,7 +39,14 @@ def __init__(self, initial_tau: float = 0.996):
...
 
-    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
+    def on_train_batch_end(self,
+                           trainer: Trainer,
+                           pl_module: LightningModule,
+                           outputs: Sequence,
+                           batch: Sequence,
+                           batch_idx: int,
+                           dataloader_idx: int
+                           ) -> None:
```
Comment (Borda) [inline pl_bolts/callbacks/byol_updates.py:49]:
```suggestion
    def on_train_batch_end(
            self,
            trainer: Trainer,
            pl_module: LightningModule,
            outputs: Sequence,
            batch: Sequence,
            batch_idx: int,
            dataloader_idx: int,
    ) -> None:
```

File: pl_bolts/callbacks/vision/confused_logit.py [hunk 10]
```diff
@@ -66,9 +69,10 @@ def __init__(
         self.logging_batch_interval = logging_batch_interval
         self.min_logit_value = min_logit_value
 
-    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
+    def on_train_batch_end(self, trainer: Trainer, pl_module: LightningModule, outputs: Sequence,
+                           batch: Sequence, batch_idx: int, dataloader_idx: int) -> None:
```
Comment (akihironitta) [inline pl_bolts/callbacks/vision/confused_logit.py:73]:
```suggestion
    def on_train_batch_end(
            self,
            trainer: Trainer,
            pl_module: LightningModule,
            outputs: Sequence,
            batch: Sequence,
            batch_idx: int,
            dataloader_idx: int
    ) -> None:
```

@Borda Shall we introduce a code formatter? [Optuna](https://github.com/optuna/optuna) uses [`black`](https://github.com/psf/black) to keep the codebase consistent in formatting, and it worked great without any issues when I previously made contributions there.

File: pl_bolts/callbacks/vision/confused_logit.py [hunk 11]
```diff
@@ -1,14 +1,17 @@
...
 import torch
-from pytorch_lightning import Callback
-from torch import nn
+from pytorch_lightning import Callback, LightningModule, Trainer
+from torch import Tensor, nn
 
 from pl_bolts.utils.warnings import warn_missing_pkg
 
-_MATPLOTLIB_AVAILABLE = importlib.util.find_spec("matplotlib") is not None
+_MATPLOTLIB_AVAILABLE = importlib.util.find_spec("matplotlib") is not None  # type: ignore[attr-defined]
```
Comment (akihironitta) [inline pl_bolts/callbacks/vision/confused_logit.py:10]: Do we need `type: ignore` here? As I ran `mypy==0.790` without it, it raises no error...

**Review by akihironitta — COMMENTED** [review]
@ydcjeff It seems there are a certain number of ignored lines, and my concern is that it can lead to ignoration of future prospective errors. Do you think we can reduce `type: ignore` from the current code? For example, in #455 by @hassiahk, we changed some lines to pass `mypy` check. https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/455#pullrequestreview-554381634

@Borda Do you think ignoring quite a few lines is fine for now?

File: pl_bolts/callbacks/vision/confused_logit.py (hunk 11 above)
Comment (ydcjeff) [reply pl_bolts/callbacks/vision/confused_logit.py:10]: Yes, it ran well locally, but failed in https://github.com/PyTorchLightning/pytorch-lightning-bolts/runs/1585245200
not sure why

**Comment by ydcjeff** [comment]
@akihironitta I am also aware of that. But most of them are from coming from torch. some are re-assigning the same variable with different types. we can probably reduce the amount of `type: ignore` in `dict_to_table`

File: pl_bolts/callbacks/vision/confused_logit.py (hunk 11 above)
Comment (akihironitta) [reply pl_bolts/callbacks/vision/confused_logit.py:10]: Ok, I see. That's strange...

File: pl_bolts/callbacks/vision/confused_logit.py [hunk 12]
```diff
@@ -1,14 +1,17 @@
...
```
Comment (akihironitta) [inline pl_bolts/callbacks/vision/confused_logit.py:10]: We actually have similar variables for other optional packages in `pl_bolts/utils/__init__.py`:
https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/a72766c461a498ab67b7c2efe7e0d49224290837/pl_bolts/utils/__init__.py#L6-L10
So, could you move the definition of `_MATPLOTLIB_AVAILABLE` to the linked file and import the variable here?

**Review by akihironitta — CHANGES_REQUESTED** [review]

**Comment by akihironitta** [comment]
> @akihironitta I am also aware of that. But most of them are from coming from torch.

If you find it impossible/difficult to remove those `type: ignore`, that's, I think, fine for now :]

**Comment by Borda** [comment]
@ydcjeff mind resolve conflict and @akihironitta min re-review?

File: pl_bolts/callbacks/data_monitor.py [hunk 13]
```diff
@@ -228,7 +229,14 @@ def __init__(self, log_every_n_steps: int = None):
...
         super().__init__(log_every_n_steps=log_every_n_steps)
 
-    def on_train_batch_start(self, trainer, pl_module, batch, *args, **kwargs):
+    def on_train_batch_start(
+            self,
+            trainer: Trainer,
+            pl_module: LightningModule,
+            batch: Sequence,
+            *args: int,
+            **kwargs: int,
```
Comment (akihironitta) [inline pl_bolts/callbacks/data_monitor.py:238]: Shouldn't these be the following?:
```suggestion
            *args: Any,
            **kwargs: Any,
```
It seems that mypy doesn't raise any errors with `int` at the moment, but I guess it may produce some in the future.

**Review by akihironitta — COMMENTED** [review]
LGTM other than the below comment! Thank you for your work!

File: pl_bolts/callbacks/data_monitor.py (hunk 13 above)
Comment (ydcjeff) [reply pl_bolts/callbacks/data_monitor.py:238]: Ahh you're right. I mismatch from the `super().on_train_batch_start()`

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
the docs fail, probably we need to add `matplotlib` also to docs requirements

**Comment by ydcjeff** [comment]
I have tested locally, the docs build still fail even if installed matplotlib.
The thing is `autodoc` tracks `_MATPLOTLIB_AVAILABLE` to the `_module_available` function. so when we print `_MATPLOTLIB_AVAILABLE`, it shows ` pytorch_lightning.utilities._module_available`, so resulting import warning.
any other solution in mind?

File: pl_bolts/callbacks/vision/confused_logit.py [hunk 14]
```diff
@@ -1,14 +1,16 @@
...
+from pl_bolts.utils import _MATPLOTLIB_AVAILABLE
 from pl_bolts.utils.warnings import warn_missing_pkg
 
-_MATPLOTLIB_AVAILABLE = importlib.util.find_spec("matplotlib") is not None
 if _MATPLOTLIB_AVAILABLE:
     from matplotlib import pyplot as plt
+    from matplotlib.axes import Axes
+    from matplotlib.figure import Figure
 else:
     warn_missing_pkg("matplotlib")  # pragma: no-cover
```
Comment (akihironitta) [inline pl_bolts/callbacks/vision/confused_logit.py:16]:
```suggestion
else:  # pragma: no cover
    warn_missing_pkg("matplotlib")
    Axes = object
    Figure = object
```

**Review by akihironitta — CHANGES_REQUESTED** [review]
@ydcjeff Sorry for the delay! I looked through how other OSS projects deal with types of optional packages, and I found that, in Optuna, they assign `object` to the types (class names) when optional packages are unavailable. https://github.com/optuna/optuna/blob/392088c25d661b997c4143fd18ccdb7480ffa825/optuna/integration/pytorch_lightning.py#L4-L12

Do you think following this convention sound reasonable?

**Review by akihironitta — APPROVED** [review]
@ydcjeff lgtm. Thank you!

File: pl_bolts/callbacks/data_monitor.py [hunk 15]
```diff
@@ -228,7 +229,14 @@ def __init__(self, log_every_n_steps: int = None):
         """
         super().__init__(log_every_n_steps=log_every_n_steps)
 
-    def on_train_batch_start(self, trainer, pl_module, batch, *args, **kwargs):
+    def on_train_batch_start(
+            self,
+            trainer: Trainer,
+            pl_module: LightningModule,
+            batch: Sequence,
```
Comment (awaelchli) [inline pl_bolts/callbacks/data_monitor.py:236]: same here

File: pl_bolts/callbacks/data_monitor.py [hunk 16]
```diff
@@ -38,22 +39,22 @@ def __init__(self, log_every_n_steps: int = None):
...
 
     def on_train_batch_start(
-        self, trainer, pl_module, batch, batch_idx, dataloader_idx
-    ):
+        self, trainer: Trainer, pl_module: LightningModule, batch: Sequence, batch_idx: int, dataloader_idx: int
+    ) -> None:
         self._train_batch_idx = batch_idx
 
-    def log_histograms(self, batch, group="") -> None:
+    def log_histograms(self, batch: Sequence, group: str = "") -> None:
```
Comment (awaelchli) [inline pl_bolts/callbacks/data_monitor.py:57]: same here

File: pl_bolts/callbacks/data_monitor.py [hunk 17]
```diff
@@ -38,22 +39,22 @@ def __init__(self, log_every_n_steps: int = None):
...
+    def on_train_start(self, trainer: Trainer, pl_module: LightningModule) -> None:
         self._log = self._is_logger_available(trainer.logger)
-        self._log_every_n_steps = self._log_every_n_steps or trainer.log_every_n_steps
+        self._log_every_n_steps = self._log_every_n_steps or trainer.log_every_n_steps  # type: ignore[attr-defined]
         self._trainer = trainer
 
     def on_train_batch_start(
-        self, trainer, pl_module, batch, batch_idx, dataloader_idx
-    ):
+        self, trainer: Trainer, pl_module: LightningModule, batch: Sequence, batch_idx: int, dataloader_idx: int
```
Comment (awaelchli) [inline pl_bolts/callbacks/data_monitor.py:53]: for batch it is too strict, this could be almost anything (Any)

**Comment by Borda** [comment]
@ydcjeff mind send another PR addressing @awaelchli comments?
